### PR TITLE
fix #8846

### LINKED ISSUE #8846: [major] `someObject is object|int|tuple` fails (but `someObject is int|object|tuple` works...)
/cc @araq
reduced from a real use case:
```nim
type ObjectLike4* = tuple|object|int
type ObjectLike5* = int|tuple|object
type ObjectLike6* = object|int|tuple

type A = object
  tint: int

echo A is ObjectLike4
echo A is ObjectLike5
echo A is ObjectLike6

echo (1,2) is ObjectLike4
echo (1,2) is ObjectLike5
echo (1,2) is ObjectLike6

echo () is ObjectLike4
echo () is ObjectLike5
echo () is ObjectLike6

var a:A
echo a is ObjectLike4
echo a is ObjectLike5
echo a is ObjectLike6
```

prints:
true
true
false
false
true
true
true
true
true
true
true
false

every entry marked as false is a bug.

looks like:
* when T is object or tuple, `T is Foo1|Foo2|Foo3` doesn't take into account the 1st element Foo`
* when T is int, no bug
* when T is an instance of a tuple, there is a bug IF the instance is of size > 0

seems like a pretty serious bug...

## probably related
```nim
#BUG:D20180902T180936:here
type ObjectLike = object|tuple
when defined(case1):
  # Error: type mismatch: got <A>
  proc fun*[T:ObjectLike](a:T)=discard
when defined(case2):
  # ok
  proc fun*[T:tuple|object](a:T)=discard

type A = object
fun(A())
```

### PR DESCRIPTION
**Pull request by metagn**
closes #8846 by making the syntax an error

### CODE REVIEW

**Comment by Araq** [comment]
Sorry, this adds more cruft to the existing cruft. Where does the grammar reflect this special casing? Inside a `type` section `object` should mean object declaration. If that means we cannot support all keyword-based type-classes via "|", so be it.

**Comment by metagn** [comment]
Makes sense, `enum` fails with this syntax anyway and the workaround is just using parentheses. I don't think there's any avoiding special casing object syntax from the rest of the language though. I think it should be fine now

**Comment by arnetheduck** [comment]
well, we've run into lots of paper cut bugs like this when attempting to use type classes which makes us lose confidence in them quite thoroughly - ie the order of the thing should not matter: either the syntax is allowed or it's not - basically, if placing `object` first moves the grammar into the `object` section, the other cases should really give an error.

Now, given that fixing is is breaking, at the very least it would be good with a compiler warning for the surprising parse.

**Comment by metagn** [comment]
@arnetheduck `type Foo = object | int` always created an invalid state though, I don't think code could have ever relied on it for it to be breaking. `type Foo = (object | int)` has always worked, so has `object | int` outside type sections. It's not an issue with typeclasses, but I can add the parentheses version as a test for typeclasses.

Also `type Foo = tuple` means `type Foo = tuple[]` and not `type Foo = (tuple)`. Even though this isn't great behavior it's consistent with Nim's grammar. I think it's too late to make any meaningful changes on this end. If Nim required `object:` with a colon from the start maybe there would be less problems, but adding this syntax now wouldn't solve any existing problems as we can't revert the old syntax.

**Comment by arnetheduck** [comment]
```
type Foo = object | int
type Xxx = object

echo Xxx is Foo
```

> created an invalid state though

how do you mean? the above prints false, which obviously is misleading and confusing.

**Comment by metagn** [comment]
> how do you mean? the above prints false, which obviously is misleading and confusing.

This:

```nim
type Foo = object | int
```

Currently means the same as:

```nim
type
  FooTemp = object
  Foo = FooTemp | int
```

except you don't have access to `FooTemp`. The invalid part is that an inaccessible nominal object is not supposed to exist, but yes it does compile and give a bad result. There is no way to rely on this that I can see, so it should error in case anyone assumed it worked correctly.

**Comment by arnetheduck** [comment]
> it should error in case anyone assumed it worked correctly.

probably want to start with a warning though?

File: compiler/parser.nim [hunk 1]
```diff
@@ -914,7 +914,8 @@ proc simpleExprAux(p: var Parser, limit: int, mode: PrimaryMode): PNode =
     pragmaExp.add result
     pragmaExp.add p.parsePragma
     result = pragmaExp
-  result = parseOperators(p, result, limit, mode)
+  if not (mode == pmTypeDef and result.kind in {nkObjectTy, nkTupleTy}):
+    result = parseOperators(p, result, limit, mode)
```
Comment (Araq) [inline compiler/parser.nim:918]: Pre-existing but this `mode` parsing is not expressible via the grammar and hence very bad.

**Comment by Araq** [comment]
Maybe the type-classes that conflict with keywords should always be required to be written in backticks, if used as a type class? For example:

```nim

type Foo = `object` | int

```

Seems a most natural solution. The old way can be deprecated and remain supported for a long time, of course.

**Comment by metagn** [comment]
> ```nim
> type Foo = `object` | int
> ```

I think this only makes sense if ``type `object`* = (object)`` is defined in system.

The breaking version of this is better done in #19383. So if it's preferred, I can revert this PR to the version that made the broken syntax work and it can be backported.

**Comment by metagn** [comment]
Closing as rejected

**Pull request closed without merging**